harness/determined/core/_checkpoint.py

```diff
@@ -133,29 +384,40 @@ def get_metadata(self, storage_id: str) -> Dict[str, Any]:
 
     @contextlib.contextmanager
     def store_path(
-        self, metadata: Optional[Dict[str, Any]] = None
+        self, metadata: Optional[Dict[str, Any]] = None, *, shard: bool = False
     ) -> Iterator[Tuple[pathlib.Path, str]]:
         """
         ``store_path()`` is a context manager which chooses a random path and prepares a directory
         you should save your model to.  When the context manager exits, the model is automatically
         uploaded (at least, for cloud-backed checkpoint storage backends).
 
-        Note that with multiple workers, only the chief worker (``distributed.rank==0``) is allowed
-        to call ``store_path()``.
+        When ``shard=False``, only the chief worker (``distributed.rank==0``) may call
+        ``store_path()``.
+
+        When ``shard=True``, ``store_path()`` becomes a synchronization point between workers, so
+        all workers must call store_path(), even workers which will not write any checkpoint files.
 
         Example:
 
         .. code::
 
-           with core_context.checkpoint.store_path() as (path, storage_id):
-               my_save_model(my_model, path)
-               print(f"done saving checkpoint {storage_id}")
-           print(f"done uploading checkpoint {storage_id}")
+           if core_context.distributed.rank == 0:
+               with core_context.checkpoint.store_path() as (path, storage_id):
```

Comment:
we need metadata here with `steps_completed` here right?  can we make sure to add it to this example and also add an example to `upload` with the metadata filled in as well?

Reply:
technically we don't since metadata is Optional in store_path and upload (not to mention `steps_completed`). 
while I am not opposed to adding metadata to the example, since this is a good practice, imo it could create confusion as to what is metadata and why it is expected. but if you feel strongly about it let me know.

btw, one thing that is for sure missing here is `shard=False`. will add it!